cpp_common/CMakeLists.txt

```diff
@@ -32,13 +34,24 @@ if(HAVE_GLIBC_BACKTRACE)
 endif(HAVE_GLIBC_BACKTRACE)
 
 add_library(${PROJECT_NAME} src/debug.cpp src/header.cpp)
-target_link_libraries(${PROJECT_NAME} ${Boost_LIBRARIES} ${console_bridge_LIBRARIES})
+target_link_libraries(${PROJECT_NAME} Boost::boost Boost::system Boost::thread ros::console_bridge)
+target_include_directories(${PROJECT_NAME} PUBLIC
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+  $<INSTALL_INTERFACE:include>
+)
 
 install(TARGETS ${PROJECT_NAME}
+  EXPORT ${PROJECT_NAME}Targets
   ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
   LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
   RUNTIME DESTINATION ${CATKIN_GLOBAL_BIN_DESTINATION})
 
 install(DIRECTORY include/
   DESTINATION ${CATKIN_GLOBAL_INCLUDE_DESTINATION}
   FILES_MATCHING PATTERN "*.h")
+
+install(EXPORT ${PROJECT_NAME}Targets
```

Comment:
`install(EXPORT ...)` repeated 4 times, this is at least redundant.

Reply:
Also I expect to see the change with `configure_package_config_file` command where you will generate `*Config.cmake` file that will load `${PROJECT_NAME}Targets.cmake`.

Reply:
As an example:
* One `install(EXPORT ...)` in top level CMakeLists.txt: https://github.com/cgold-examples/fruits/blob/b266e5a8b0dccc250390346acde07e613562d85d/CMakeLists.txt#L120-L124
* `configure_package_config_file` to generate `*Config.cmake`: https://github.com/cgold-examples/fruits/blob/b266e5a8b0dccc250390346acde07e613562d85d/CMakeLists.txt#L98-L105
* `install(TARGETS ... EXPORT ...)` one for each target:
  * https://github.com/cgold-examples/fruits/blob/b266e5a8b0dccc250390346acde07e613562d85d/lib/fruits/rosaceae/CMakeLists.txt#L24-L31
  * https://github.com/cgold-examples/fruits/blob/b266e5a8b0dccc250390346acde07e613562d85d/lib/fruits/tropical/CMakeLists.txt#L24-L31

Reply:
> * One install(EXPORT ...) in top level CMakeLists.txt: https://github.com/cgold-examples/fruits/blob/b266e5a8b0dccc250390346acde07e613562d85d/CMakeLists.txt#L120-L124
> * configure_package_config_file to generate *Config.cmake: https://github.com/cgold-examples/fruits/blob/b266e5a8b0dccc250390346acde07e613562d85d/CMakeLists.txt#L98-L105

I will reply in more detail to every comment later but for now I just wanted to give a brief rationale:

* Agree that it should do it. But if you notice the repository actually is a composition of several `project()`, that is why it is reapeated. Because each project is then installed as a separate package. It's a bit non-standard, but that's the way all ROS repositories have it. Bigger repos like www.github.com/ros/ros_comm can have as many as 20 sub-packages like that - so I just found this to be the only practical way how to actually hunterise the whole repository without having to hunterise each sub-package separately.
I guess perhaps another option is to remove this system completely, but I thought that would actually be more maintenance burden, because the patches would be much more extensive.

Also we need to keep the catkin_package() macro which designates the project as a `catkin` package - as this is important for any package to be integrated into tools like [roslaunch](http://wiki.ros.org/roslaunch) and others.

* `configure_package_config_file`: the catkin_package macro already does this, with a custom *Config.cmake generation: https://github.com/hunter-packages/catkin/blob/be5e2754cd32fd2d2cb0f9b5582239769f0624d2/cmake/catkin_package.cmake#L505

Reply:
> configure_package_config_file: the catkin_package macro already does this, with a custom *Config.cmake generation

Okay, I see

Reply:
Mostly I mean that if there are no plans to add this code upstream then for Hunter you can have only one at the top.